python/ray/serve/backend_state.py

```diff
@@ -1039,3 +1155,30 @@ def update(self) -> bool:
                 self._notify_replica_handles_changed(tag)
                 for tag in transitioned_backend_tags
             ]
+
+        # After observe & shuffle is done with replicas sitting at new state,
+        # determine which deployment goals succeeded or failed.
+        (complete_goal_ids, failed_goal_ids) = self._check_completed_goals()
+
+        for backend_tag, goal_id in complete_goal_ids:
+            self._replica_failed_to_start_counter[backend_tag] = -1
```

Comment:
that's one way of doing it too. It's doing it here in this revision because we makes a call when retry cap is reached and consider deployment as successful and this is where we handle goal ids that reached terminal state.

I think technically this counter can be set in different places to be considered correct, purely depends on how & where we make the call.